### ISSUE TITLE
Missing tests for sockets

### ISSUE THREAD

**Issue by RealyUniqueName** [HaxeFoundation/haxe#6889]
Currently we have a test only for `sys.net.Socket.bind()`. And even that test does not look ok: https://github.com/HaxeFoundation/haxe/blob/development/tests/sys/src/net/TestSocket.hx#L24

**Comment by ncannasse** [comment]
We're missing tests for the whole Sys API. Anyone willing to contribute
these?

On Mon, Mar 5, 2018 at 7:01 AM, Alexander Kuzmenko <notifications@github.com

**Comment by bendmorris** [comment]
I added some basics for `File` but Socket requires integration tests and would be tricky to fit into the current test model.

I'm working on Python/PHP's Socket, I'll take a shot at some basic tests to verify.

**Comment by ncannasse** [comment]
In case that helps you can do local networking opening a server +
connecting a client on 127.0.0.1 in the same process

**Comment by RealyUniqueName** [comment]
IMHO we should rethink the whole sys api. But that may be too much atm.

**Comment by RealyUniqueName** [comment]
E.g. `sys.db` looks like it was designed only for MySQL. And specifically for neko implementation of MySQL. It lacks proper error handling, retrieving column types, charsets, connection settings (at least timeout), etc. 
I'm not sure if `sys.db.Connection` could be effectively implemented for Oracle or Postgres. E.g. `Connection.lastInsertId()` is not even a thing in Oracle. And what is the purpose of `Connection.addValue(stringBuf, value)`? Why does it exist in this place? It looks like a simple utility method which has nothing to do with a connection.

`sys.io` is missing pipes, forks, process ownership info, sending signals to processes, modifying file attributes (atime, mtime etc.), managing file permissions, ownership, etc.

I have little experience with sockets. But out of the top of my head: Haxe sockets don't have `poll` implementation.

That's what I can point at just briefly looking through `sys` package.

Current sys api is fine for a lot of common tasks. It's really cross platform, and I like its simplicity. But overall it makes an impression of an incomplete and not well-thought api. I guess it's the consequence of Haxe being created and evolved as a tool mainly for game development and later for web development.
So I think this aspect of Haxe could be greatly improved. It's definitely not a top priority task. And yet I'd like to do something in this area.

**Comment by jdonaldson** [comment]
I was waiting on a good async solution for haxe to emerge.  Many haxe targets are single threaded, and will require async for non-trivial IO ops.

**Comment by ncannasse** [comment]
For async I'm wrapping libuv in HashLink so we get basically the same low
level implementation as Node.
That could apply to HxCpp and maybe Java/C# as well. I don't think Neko
will get much updates.

**Comment by jdonaldson** [comment]
What do you think is a good next step for other targets?  Should we follow the same api as hl in the uv package?

**Comment by RealyUniqueName** [comment]
I suggest to wait for coroutines implementation before designing async sys api.

**Comment by ousado** [comment]
> So I think this aspect of Haxe could be greatly improved. It's definitely not a top priority task. And yet I'd like to do something in this area.

Agreed - except for the "top priority" part. Haxe could be greatly improved here, and it could be an outstanding language for systems programming, easily on par with the main garbage collected languages in that field. Unfortunately, Haxe still doesn't even have native support for an integer type that's wide enough to reliably represent a file offset, or an inode, which is a major issue for any serious attempt at a cross-platform and cross-target sys API.

**Comment by jdonaldson** [comment]
There's a pull request for the missing sync socket api here : https://github.com/HaxeFoundation/haxe/pull/6887

I'm writing up the haxe version of the socket api for lua over here : https://github.com/HaxeFoundation/haxe/tree/python-socket-lua
I'll merge it over on that pull request first.

Writing up the code to satisfy the old sys.Socket api really underlines the need for an async version.  Sockets are not handled well by Haxe.  They're reliant on exceptions, their interface is often undefined in some situations (the input/output fields are typically not defined until the socket is bound, etc.), and the sys api exposes a lot of esoteric bits of the underlying c library.  (Quick, can anyone tell me what the "exception" argument for socket.select pertains to?)

Long story short, if we get a good async api, we can relegate sys.Socket to the legacy pile.  If we come up with a more straight forward api, I doubt anybody will miss it.

**Comment by bendmorris** [comment]
I think this also highlights a need to track code coverage. Over half of the Socket implementations were incomplete or inconsistent. How many other std APIs may have the same problem?

**Comment by ousado** [comment]
Agreed, especially the exceptions are problematic. I'd very much like to have an API layer that mirrors the BSD socket API closely, though, I think it's hard to do better without sacrificing flexibility. 

In Rust-land it's common to create an 1:1 API wrapper with a "sys" suffix, so `epoll-sys`, for example, and then add a safe, idiomatic layer on top. That allows different users on the safe layer to take advantage of the 1:1 system API wrappers, without having to do all the legwork all over again. I think that's a very reasonable approach, and we should adopt it.

**Comment by Simn** [comment]
Nothing actionable here for Haxe 4.

**Comment by Simn** [comment]
I think the consensus here is to focus on the new asys API from space and not invest any more time into the old API, so I'll go ahead and close this.
